warehouse/query-core/src/main/java/datawave/query/iterator/QueryLogIterator.java

```diff
@@ -0,0 +1,191 @@
+package datawave.query.iterator;
+
+import static datawave.query.iterator.QueryOptions.QUERY_ID;
+
+import java.io.IOException;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Map;
+
+import org.apache.accumulo.core.data.ByteSequence;
+import org.apache.accumulo.core.data.Key;
+import org.apache.accumulo.core.data.Range;
+import org.apache.accumulo.core.data.Value;
+import org.apache.accumulo.core.iterators.IteratorEnvironment;
+import org.apache.accumulo.core.iterators.OptionDescriber;
+import org.apache.accumulo.core.iterators.SortedKeyValueIterator;
+import org.apache.log4j.Logger;
+
+/**
+ * An iterator used to log the start and end of each method run by the {@link SortedKeyValueIterator} above it in the iterator stack. Logs the QueryID
+ * associated with each method. Logs are written on the TServer where the iterator is running
+ */
+public class QueryLogIterator implements SortedKeyValueIterator<Key,Value>, OptionDescriber {
+
+    private static final Logger log = Logger.getLogger(QueryLogIterator.class);
+    private static final String CLASS_NAME = QueryLogIterator.class.getSimpleName();
+
+    private String queryID;
+    private SortedKeyValueIterator<Key,Value> source;
+    private IteratorEnvironment env;
+
+    /**
+     * Default constructor
+     */
+    public QueryLogIterator() {
+        // no-arg constructor
+    }
+
+    /**
+     * Class copy constructor
+     */
+    public QueryLogIterator(QueryLogIterator other, IteratorEnvironment env) {
+        this.source = other.source.deepCopy(env);
+        this.env = other.env;
+        this.queryID = other.queryID;
+    }
+
+    /**
+     * Wraps the init() method of the iterator above it, logging the start/end of the method along with its query id.
+     */
+    @Override
+    public void init(SortedKeyValueIterator<Key,Value> source, Map<String,String> options, IteratorEnvironment env) throws IOException {
+
+        try {
+            logStartOf("init()");
+            this.queryID = options.get(QUERY_ID);
+            this.source = source;
+            this.env = env;
+        } finally {
+            logEndOf("init()");
```

Comment:
Not sure there is value in calling logStartOf and logEndOf here since we are not making any call on source between these two calls